service/controller/resource/clusterconfigmap/desired.go

```diff
@@ -67,56 +66,52 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) ([]*cor
 		clusterCA = string(secret.Data["tls.crt"])
 	}
 
-	appOperatorValues := map[string]interface{}{
-		"app": map[string]interface{}{
-			"watchNamespace":    cr.GetNamespace(),
-			"workloadClusterID": key.ClusterID(&cr),
-		},
-		"provider": map[string]interface{}{
-			"kind": r.provider,
-		},
-		"registry": map[string]interface{}{
-			"domain": r.registryDomain,
-		},
-	}
+	// clusterDNSIP contains the `coredns` k8s `Service` IP.
+	// This IP needs to belong to the `Services` CIDR configured for the k8s cluster, which can be set in the
+	// "serviceSubnet" field of the KubeadmControlPlane CR. If this field is set we want to take the IP from that CIDR.
+	// If it's not, we take the IP from the CIDR passed as parameter, which will probably be the default Service CIDR.
+	var clusterDNSIP = r.dnsIP
+	{
+		if cr.Spec.ControlPlaneRef != nil {
+			kubeadmControlPlane := &unstructured.Unstructured{}
+			kubeadmControlPlane.SetGroupVersionKind(schema.GroupVersionKind{
+				Group:   cr.Spec.ControlPlaneRef.GroupVersionKind().Group,
+				Kind:    cr.Spec.ControlPlaneRef.Kind,
+				Version: cr.Spec.ControlPlaneRef.GroupVersionKind().Version,
+			})
+			err = r.k8sClient.CtrlClient().Get(ctx, client.ObjectKey{
+				Namespace: cr.Namespace,
+				Name:      cr.Spec.ControlPlaneRef.Name,
+			}, kubeadmControlPlane)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
 
-	clusterValues := map[string]interface{}{
-		"baseDomain": key.BaseDomain(&cr, r.baseDomain),
-		"chartOperator": map[string]interface{}{
-			"cni": map[string]interface{}{
-				"install": true,
-			},
-		},
-		"cluster": map[string]interface{}{
-			"calico": map[string]interface{}{
-				"CIDR": podCIDR,
-			},
-			"kubernetes": map[string]interface{}{
-				"API": map[string]interface{}{
-					"clusterIPRange": r.clusterIPRange,
-				},
-				"DNS": map[string]interface{}{
-					"IP": r.dnsIP,
-				},
-			},
-		},
-		"clusterCA":    clusterCA,
-		"clusterDNSIP": r.dnsIP,
-		"clusterID":    key.ClusterID(&cr),
-		"clusterCIDR":  "",
-		"provider":     "unknown",
+			serviceCidr, serviceCidrFound, err := unstructured.NestedString(kubeadmControlPlane.Object, []string{"spec", "kubeadmConfigSpec", "clusterConfiguration", "networking", "serviceSubnet"}...)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+
+			if serviceCidrFound {
+				clusterDNSIP, err = key.DNSIP(serviceCidr)
+				if err != nil {
+					return nil, microerror.Mask(err)
+				}
+			}
+		}
 	}
 
+	var (
+		// clusterCIDR is only used on azure.
+		clusterCIDR = ""
+		// gcpProject is only used on gcp.
+		gcpProject = ""
+	)
 	{
 		infrastructureRef := cr.Spec.InfrastructureRef
 		if infrastructureRef != nil {
-			switch infrastructureRef.Kind {
-			case "AWSCluster":
-				clusterValues["provider"] = "aws"
-
-			case "AzureCluster":
-				clusterValues["provider"] = "azure"
-
+			switch r.provider {
+			case "azure":
 				var azureCluster capz.AzureCluster
```

Comment:
as follow up task i would recommend to use the unstructured client for all infraspecific implementation. It's already used for GCP and we should continue on that for other providers.

But i won't block this right now.

Reply:
Yeah, agree. But I believe this logic should be removed in the future, if we create the `cluster-values` configmap from the `cluster-*` apps. So I will leave it like that for now.